enhancements/update/synchronized-upgrades.md

```diff
@@ -0,0 +1,302 @@
+
+Thus, other than ovn-kubernetes, no OCP component needs to be
+concerned about version skew between the two clusters, because the
+components in the two clusters are completely unaware of each other.
+
+For ovn-kubernetes, if we ever change the details of the cross-cluster
+communication, then we will need to add proper checks to enforce
+tolerable cross-cluster skew at that time.
+
+## Design Details
+
+The upgrade plan takes advantage of two things:
+
+- Other than the reboots, no part of the upgrade needs to be
+  synchronized between the two clusters.
+
+- Each infra node only runs pods to support the pods on its
+  corresponding tenant node. Thus, if a tenant node reboots, it is
+  safe to immediately reboot its infra node as well without needing
+  to cordon and drain it, because it is guaranteed that none of its
+  pods are doing anything important at that point.
+
+So, both clusters' upgrades can proceed normally up until the MCO
+upgrade.
+
+In the infra cluster, rather than cordoning, draining, and rebooting
+nodes one-by-one, the MCO will instead queue up the new RHCOS image on
+each node, and then mark the node (somehow) as "waiting for a reboot".
+Then the MCO will watch the nodes and wait for each one to reboot (of
+its own accord), and then finally mark itself fully-upgraded after
+that happens.
+
+On the infra nodes, when they are "waiting for a reboot", they will
+monitor the state of the ovn-kubernetes network; when they determine
+that the tenant node is rebooting, the infra node will then also
+reboot.
+
+In terms of synchronization, this means that we want to avoid the
+tenant cluster reaching the MCO upgrade stage when the infra cluster
+is not ready for the tenant cluster to start rebooting nodes:
+
+- One way to do this would be to not let a tenant cluster start an
+  upgrade until the infra cluster has already reached the "waiting
+  for reboots" stage. This could be done by having an operator in
+  the tenant cluster whose status is always `Upgradeable: False`
+  except when the infra cluster tells it that it is safe to
+  proceed.
+
+- Another possibility would be to have an operator that sits just
+  before the MCO in the upgrade order, which checks if the infra
+  cluster is in the middle of an upgrade, and if so, blocks until
+  the infra cluster is ready for the tenant cluster to proceed. (Or
+  rather, the infra cluster would notify it that it should/shouldn't
+  block the upgrade.)
+
+  - This would allow for the possibility of doing tenant cluster
+    upgrades without corresponding infra cluster upgrades, which
+    might be useful?
+
+### Open Questions
+
+- How exactly will the coordination between the two MCO upgrades
+  occur?
+
+- How exactly will infra nodes detect their tenant node rebooting?
+
+- Do we want to allow (z-stream) upgrades of the tenant cluster
+  without also upgrading the infra cluster?
+
+### Risks and Mitigations
+
+The infra cluster could remain stuck at the pre-reboot stage of the
```

Comment:
The prereboot stage means that the upgrade has been performed except for the final reboot, this doesn't seem like a state we want all nodes to stay in at all?

Reply:
> Also what happens if the update is bad? All nodes would be scheduled to make the update as opposed to a rolling mechanism that we have now.

Yeah... so if you look at [the first draft of the enhancement](https://github.com/openshift/enhancements/commit/8033a9236774dc115633adebb2610a8fee76599f), I had suggested a more complicated system, where the MCOs of the two clusters would agree on a particular order to upgrade nodes in, and then each individual node upgrade would be synchronized. Scott suggested that we didn't actually need that much synchronization and I rewrote the enhancement to the current form. But maybe that won't work...